src/gen-mapping.ts

```diff
@@ -336,6 +355,113 @@ export class GenMapping {
     };
+
+    applySourceMap = (
+      map: GenMapping,
+      consumer: TraceMap,
+      rawSourceFile?: string | null,
+      sourceMapPath?: string | null,
+    ) => {
+      let sourceFile = rawSourceFile ?? consumer.file;
+      if (sourceFile == null) {
+        throw new Error(
+          'applySourceMap requires either an explicit source file, ' +
+            'or the source map\'s "file" property. Both were omitted.',
+        );
+      }
+
+      if (!sourceMapPath) sourceMapPath = '';
+      else if (!sourceMapPath.endsWith('/')) sourceMapPath += '/';
+
+      const { _mappings: mappings, _sourcesContent: sourcesContent, sourceRoot } = map;
+      const sources = map._sources.array;
+      const names = map._names.array;
+
+      const {
+        sources: consumerSources,
+        sourcesContent: consumerSourcesContent,
+        names: consumerNames,
+      } = consumer;
+
+      let sourceIndex = get(map._sources, sourceFile);
+      if (sourceIndex === undefined && sourceRoot) {
+        // If we couldn't fine the source file and there's a sourceRoot, the sourceFile may have
+        // been joined with the root already. Try again after making the file relative to the root.
+        sourceFile = relativeUri(sourceRoot, sourceFile);
+        sourceIndex = get(map._sources, sourceFile);
+      }
+
+      // If we still can't find the source file, then there's no way we could remap the file with
+      // the new consumer map.
+      if (sourceIndex === undefined) return;
+
+      // The source file and several names could be replaced by the remapped consumer map. To avoid
+      // leaving dead entires, we regenerate the both (and the paired sourcesContent).
+      const newSources = new SetArray();
+      const newSourcesContent: (string | null)[] = [];
+      const newNames = new SetArray();
+
+      for (let i = 0; i < mappings.length; i++) {
+        const line = mappings[i];
+
+        for (let j = 0; j < line.length; j++) {
+          const seg = line[j];
+          if (seg.length === 1) continue;
+
+          let traced = null;
+          if (seg[1] === sourceIndex) {
+            traced = traceSegment(consumer, seg[2], seg[3]);
+          }
+
+          if (traced == null) {
+            // Either this segment is mapping a different source file, or we couldn't traced the
+            // original mapping in the consumer. Either way, we'll keep the original.
+            const initialIndex = seg[1];
+            const newSourceIndex = (seg[1] = put(newSources, sources[initialIndex]));
+            newSourcesContent[newSourceIndex] = sourcesContent[initialIndex];
+            if (seg.length === 5) seg[4] = put(newNames, names[seg[4]]);
+            continue;
+          }
+
+          if (traced.length === 1) {
+            // If the traced mapping points to a sourceless segment, we need to truncate the
+            // original to also be sourceless.
+            (seg as number[]).length = 1;
+            continue;
+          }
```

Comment:
This is the last block that isn't covered by tests.
Since this was introduced by your PR to my PR, I have to admit I'm not sure what case it covers and what input would cover this test. do you have an example ?